hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -212,16 +211,24 @@ type SymbolProvider = Uri -> IdeDeferM (IdeResult [DocumentSymbol])
 data FormattingType = FormatDocument
                     | FormatRange Range
 
--- | Formats the given Uri with the given options.
+-- | Formats the given Text associated with the given Uri.
+-- Should, but might not, honor the provided formatting options (e.g. Floskell does not).
 -- A formatting type can be given to either format the whole document or only a Range.
+-- 
+-- Text to format, may or may not, originate from the associated Uri. 
+-- E.g. it is ok, to modify the text and then reformat it through this API.
+--
+-- The Uri is mainly used to discover formatting configurations in the file's path.
+--
 -- Fails if the formatter can not parse the source.
--- Failing menas here that a IdeResultFail is returned.
+-- Failing means here that a IdeResultFail is returned.
 -- This can be used to display errors to the user, unless the error is an Internal one.
 -- The record 'IdeError' and 'IdeErrorCode' can be used to determine the type of error.
-type FormattingProvider = Uri -- ^ Uri to the file to format. Can be mapped to a file with `pluginGetFile`
+type FormattingProvider = T.Text -- ^ Text to format
+        -> Uri -- ^ Uri of the file being formatted
         -> FormattingType  -- ^ How much to format
         -> FormattingOptions -- ^ Options for the formatter
-        -> IdeDeferM (IdeResult [TextEdit]) -- ^ Result of the formatting or the unchanged text.
+        -> IdeM (IdeResult [TextEdit]) -- ^ Result of the formatting or the unchanged text.
```

Comment:
This used to be `IdeDeferM` because brittany needs the GHC parsed source to work. Now that it is `IdeM`, will be potentially lose formatting under some circumstances?

Reply:
Brittany's `FormattingProvider` doesn't use `ifCachedModule/withCachedModule` on master at the moment, so it looks like it wouldn't have been waiting for the parsed source
https://github.com/haskell/haskell-ide-engine/blob/a48a02cbd7efdf5ddc923ab7458921398a8c29cd/src/Haskell/Ide/Engine/Plugin/Brittany.hs#L42-L64

It looks like I moved it from `IdeGhcM` to `IdeM` in this [commit](https://github.com/haskell/haskell-ide-engine/commit/d39b4eec4fd0c8e4fcb406b425f98f4a3dda0e3a), and in retrospect I should have given a second thought as to why it was In `IdeGhcM` in the first place. 
`runBrittany` just runs in `IO`: does it access the same ghc-mod instance as HIE?

Reply:
> runBrittany just runs in IO: does it access the same ghc-mod instance as HIE?

Good question. I think once my HaRe update to use the new hie-bios is done we will have a clearer view of the API, and can ask brittany to expose something that makes use of the module caching

If it does not, then IdeDeferM makes no difference.